### ISSUE TITLE
Provide a way to override configs

### ISSUE THREAD

**Issue by edg2s** [MatmaRex/patchdemo#19]
This would allow demonstrations of features hidden behind feature flags, for example.

We don't want to let people add arbitrary PHP to local settings, so we'd probably need to do something JSON-based instead.

**Comment by DemianX0** [comment]
@edg2s Do you need to change the flags after the demo was deployed or the flags could be part of the patch?

**Comment by edg2s** [comment]
How would that look? It would be much better to not have to pollute the patch with unnecessary config changes.

**Comment by DemianX0** [comment]
I agree that's not very user-friendly. I imagine adding an extra patch to the patchdemo, including only the configuration. That would save the important patch(es) from pollution and only add the price of one extra commit. The benefit: zero complications with another settings file, security, etc.

**Comment by DemianX0** [comment]
@edg2s An example:  https://gerrit.wikimedia.org/r/c/mediawiki/core/+/585856

There are alternative approaches, ex: Patchdemo could generate an optional import at the end of LocalSettings.php for PatchDemoSettings.php

Opinions?

**Comment by edg2s** [comment]
I'd rather we didn't pollute Gerrit with these patches. It's also more of a burden on the developer, they will need to have this cherry pick a config patch to every patch they want to test, and part of the purpose of this tool is to allow non-devs (e.g. QA's) to test patches as well.

**Comment by DemianX0** [comment]
> I'd rather we didn't pollute Gerrit with these patches.
Me too. Until there's a better solution though...

> It's also more of a burden on the developer, they will need to have this cherry pick a config patch to every patch they want to test

No, just apply two patches with patchdemo. See [my last 2 demos](https://patchdemo.wmflabs.org) (list will change with time).

**Comment by edg2s** [comment]
> Me too. Until there's a better solution though...

This task is about coming up with a better solution.

**Comment by jdlrobson** [comment]
Maybe a JSON of config values could be used to generate a LocalSettings.php?

e.g. LocalSettings.json
it would avoid arbitary PHP being run at least.

```
{ "Foo": 4, "stuff": [] }
```
would generate
```
$wgFoo = 4;
$wgstuff = [];
```

**Comment by DemianX0** [comment]
Implicit prefixing has many pitfalls, it's safer to use full variable names with "wg" prefix constraint:
```
{ "wgFoo": 4, "wgstuff": [] }
```

Regarding the format: accept strict yaml in addition to json. Everyone has their opinion of both json and yaml (good or bad or both)... In any case either one or the other is more suitable for a given task or development approach. The developer should have the choice, not being forced to use one.

**Comment by edg2s** [comment]
JSON decoding comes out of the box with PHP unlike YAML, and in general we shouldn't add complexity to support developer preferences.

**Comment by jdlrobson** [comment]
The implicit prefix would intentionally restrict what you can do in this file. My concern is arbitrary PHP being added so any safeguard helps with that.

**Comment by DemianX0** [comment]
Yes. Enforcing (constraining) the prefix is just as good imo.

**Comment by DemianX0** [comment]
> JSON decoding comes out of the box with PHP unlike YAML, and in general we shouldn't add complexity to support developer preferences.

Now that the Yaml parser is [pulled into this tool](https://github.com/MatmaRex/patchdemo/commit/ed77121b3cf203ba00b1f4ff8a169e1b361dddc4#diff-90696b34b937136cdebf9c08aeb14d23R114), both config.json and config.yaml can be supported as well.

I personally don't like either json or yaml (read as: :nauseated_face:), and prefer HOCON or JSON5. I guess those are not on the menu, so for the least pain, please support both popular candidates :wink:
